File: day16/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>
#include <sstream>
#include <unordered_set>
#include <unordered_map>
#include <map>
#include <list>
#include <regex>
#include <variant>
#include <vector>
#include <future>

auto read() {
    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");
    std::string line;
    std::vector<std::string> result;
    while (std::getline(ifs, line)) {
        if (line.empty())
            continue;
        result.push_back(line);
    }
    return result;
}

auto read_all() {
    std::fstream ifs(std::string(SRC_DIR) + "/input.txt");
    std::stringstream data;
    data << ifs.rdbuf();
    return data.str();
}

struct Node {
    int rate = 0;
    std::string from;
    std::vector<std::string> to;
};

Node parse(const std::string &data) {

    //Valve AA has flow rate=0; tunnels lead to valves DD, II, BB
    static std::regex base_regex("Valve ([A-Z]+) has flow rate=([0-9]+); tunnels? leads? to valves? ([A-Z ,]+)");
    static std::regex valve_regex("([A-Z]+)");

    std::smatch base_match;
    auto res = std::regex_match(data, base_match, base_regex);
    assert(res);
    Node node;
    node.from = base_match[1];
    node.rate = std::stoi(base_match[2]);
    auto valves = base_match[3].str();
    for (auto it = std::sregex_iterator(valves.begin(), valves.end(), valve_regex);
         it != std::sregex_iterator(); ++it) {
        node.to.push_back(it->str());
    }
    return node;
}

template<class T>
inline void hash_combine(std::size_t &seed, const T &v) {
    std::hash<T> hasher;
    seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
}

template<class A, class B>
struct std::hash<std::pair<A, B>> {
    size_t operator()(const std::pair<A, B> &pair) const {
        std::size_t seed = 0;
        hash_combine(seed, pair.first);
        hash_combine(seed, pair.second);
        return seed;
    }
};

namespace std {
    template<size_t N>
    struct less<bitset<N> > : binary_function<bitset<N>, bitset<N>, bool> {
        bool operator()(const bitset<N> &L, const bitset<N> &R) const {
            return L.to_ulong() < R.to_ulong();
        }
    };
}

bool operator<(const std::bitset<16> &l, const std::bitset<16> &r) {
    return l.to_ulong() < r.to_ulong();
}

struct State {
    int node = 0;
    int64_t score = 0;
    std::bitset<16> opened;
};

int64_t bfs(const std::vector<std::vector<int>> &graph, const std::vector<int> &rates, int start, int minutes,
            const std::bitset<16> &opened) {
    int64_t result = 0;
    std::unordered_map<std::pair<std::bitset<16>, int64_t>, int64_t> memo;
    std::deque<State> q = {{start, 0, opened}};
    for (int i = minutes; i > 0; --i) {
        int size = q.size();
        while (size-- > 0) {
            const auto cur = q.front();
            q.pop_front();

            std::pair<std::bitset<16>, int> key = {cur.opened, cur.node};
            auto it = memo.find(key);
            if (it != memo.end() && it->second >= cur.score)
                continue;

            memo[key] = cur.score;

            result = std::max(result, cur.score);

            if (!cur.opened[cur.node] && rates[cur.node] > 0) {
                auto next = cur;
                next.opened[cur.node] = true;
                next.score += (rates[cur.node] * (i - 1));
                q.push_back(next);
            }

            for (int next: graph[cur.node]) {
                q.push_back({next, cur.score, cur.opened});
            }
        }
    }
    return result;
}

struct IndexGen {
    auto operator()(const std::string &name) {
        auto it = indexes.find(name);
        if (it != indexes.end())
            return it->second;
        return indexes[name] = idx++;
    };

    int size() {
        return idx;
    }

private:
    int idx = 0;
    std::map<std::string, int> indexes;
};

void part1() {

    IndexGen index;

    std::vector<int> rates(1000);
    std::vector<std::vector<int>> graph(1000);
    for (const auto &line: read()) {
        Node node = parse(line);
        rates[index(node.from)] = node.rate;
        for (const auto &elem: node.to)
            graph[index(node.from)].push_back(index(elem));
    }
    rates.resize(index.size());
    graph.resize(index.size());

    auto result = bfs(graph, rates, index("AA"), 30, {});
    std::cout << "Part1: " << result << std::endl;
}

std::map<std::bitset<16>, int64_t>
find_pairs(const std::vector<std::vector<int>> &graph, const std::vector<int> &rates, int start, int minutes) {
    std::map<std::bitset<16>, int64_t> result;
    std::unordered_map<std::pair<std::bitset<16>, int64_t>, int64_t> memo;
    std::deque<State> q = {{start, 0, {}}};
    for (int i = minutes; i > 0; --i) {
        int size = q.size();
        while (size-- > 0) {
            const auto cur = q.front();
            q.pop_front();

            std::pair<std::bitset<16>, int> key = {cur.opened, cur.node};
            auto it = memo.find(key);
            if (it != memo.end() && it->second >= cur.score)
                continue;

            memo[key] = cur.score;

            result[cur.opened] = std::max(result[cur.opened], cur.score);

            if (!cur.opened[cur.node] && rates[cur.node] > 0) {
                auto next = cur;
                next.opened[cur.node] = true;
                next.score += (rates[cur.node] * (i - 1));
                q.push_back(next);
            }

            for (int next: graph[cur.node]) {
                q.push_back({next, cur.score, cur.opened});
            }
        }
    }
    return result;
}


void part2() {

    IndexGen index;
    std::vector<int> rates(1000);
    std::vector<std::vector<int>> graph(1000);
    for (const auto &line: read()) {
        Node node = parse(line);
        rates[index(node.from)] = node.rate;
        for (const auto &elem: node.to)
            graph[index(node.from)].push_back(index(elem));
    }
    rates.resize(index.size());
    graph.resize(index.size());

    auto start_idx = index("AA");
    auto possible_comps = find_pairs(graph, rates, start_idx, 26);

    int64_t result = 0;
    for (const auto &elem: possible_comps) {
        auto elephant = bfs(graph, rates, start_idx, 26, elem.first);
        result = std::max(result, elem.second + elephant);
    }

    std::cout << "Part2: " << result << std::endl;
}

int main() {
    part1();
    part2();
    return 0;
}

File: day16/src/main.rs
#![allow(unused_imports)]

use std::cell::RefCell;
use std::cmp::Ordering;
use std::collections::{HashMap, VecDeque, HashSet};
use std::fs::File;
use std::io::{self, BufRead};
use std::ops::Deref;
use std::path::Path;
use std::rc::Rc;
use std::vec;
use regex::{Captures, Regex};
use scanf::sscanf;
use lazy_static::lazy_static;

fn read_all() -> String {
    let file_path = "input.txt";
    std::fs::read_to_string(file_path).unwrap()
}

fn read_lines() -> Vec<String> {
    read_all().split('\n').map(|x| x.parse().unwrap()).filter(|x: &String| !x.is_empty()).collect()
}

struct Node {
    rate: i64,
    from: String,
    tos: Vec<String>,
}

fn parse(input: &str) -> Node {
    lazy_static! {
        static ref BASE_RE: Regex = Regex::new(r#"Valve (?P<from>[A-Z]+) has flow rate=(?P<rate>[0-9]+); tunnels? leads? to valves? (?P<tos>[A-Z ,]+)"#).unwrap();
        static ref VALVE_RE: Regex = Regex::new(r#"(?P<to>[A-Z]+)"#).unwrap();
    }

    match BASE_RE.captures(input) {
        Some(cap) => {
            return Node {
                rate: cap.name("rate").unwrap().parse::<i64>().unwrap(),
                from: cap.name("from").unwrap().to_string(),
                tos: VALVE_RE.captures_iter(cap.name("tos").unwrap()).map(|x| x.name("to").unwrap().to_string()).collect(),
            };
        }
        _ => panic!("no regex match")
    }
}

struct State {
    node: i64,
    score: i64,
    opened: [bool; 64],
}

fn bfs(graph: &HashMap<i64, (i64, Vec<i64>)>, start: i64, minutes: i64, opened: [bool; 64]) -> i64 {
    let mut q: VecDeque<State> = VecDeque::new();
    q.push_back(State { node: start, score: 0, opened });

    let mut memo: HashMap<(i64, [bool; 64]), i64> = HashMap::new();

    let mut result = 0i64;
    for minute in (1..=minutes).rev() {
        let qsize = q.len();
        for _ in 0..qsize {
            let cur = q.pop_front().unwrap();

            if let Some(&cached) = memo.get(&(cur.node, cur.opened)) {
                if cached >= cur.score {
                    continue;
                }
            }
            memo.insert((cur.node, cur.opened), cur.score);

            result = result.max(cur.score);

            if !cur.opened[cur.node as usize] {
                let node = graph.get(&cur.node).unwrap();
                if node.0 > 0 {
                    let mut opened = cur.opened.clone();
                    opened[cur.node as usize] = true;
                    q.push_back(State { node: cur.node, score: &cur.score + (node.0 * (minute - 1)), opened });
                }
            }

            for next in &graph.get(&cur.node).unwrap().1 {
                q.push_back(State { node: *next, score: cur.score, opened: cur.opened });
            }
        }
    }
    result
}

struct Index {
    size: i64,
    map: HashMap<String, i64>,
}

impl Index {
    fn new() -> Index {
        Index {
            size: 0,
            map: HashMap::new(),
        }
    }

    fn get(&mut self, val: String) -> i64 {
        if let Some(x) = self.map.get(&val) {
            return *x;
        }
        let newidx = self.size;
        self.map.insert(val, newidx);
        self.size += 1;
        return newidx;
    }
}

fn part1() {
    let mut index = Index::new();
    let mut graph: HashMap<i64, (i64, Vec<i64>)> = HashMap::new();

    {
        let nodes: Vec<Node> = read_lines().iter().map(|x| parse(x)).collect();
        for node in nodes {
            graph.insert(index.get(node.from), (node.rate, node.tos.iter().map(|x| index.get(x.clone())).collect()));
        }
    }

    let result = bfs(&graph, index.get("AA".to_string()), 30, [false; 64]);
    println!("Part1: {}", result);
}

fn find_pairs(graph: &HashMap<i64, (i64, Vec<i64>)>, start: i64, minutes: i64) -> HashMap<[bool;64], i64> {
    let mut result = HashMap::new();

    let mut q: VecDeque<State> = VecDeque::new();
    q.push_back(State { node: start, score: 0, opened: [false;64] });

    let mut memo: HashMap<(i64, [bool; 64]), i64> = HashMap::new();

    for minute in (1..=minutes).rev() {
        let qsize = q.len();
        for _ in 0..qsize {
            let cur = q.pop_front().unwrap();

            if let Some(&cached) = memo.get(&(cur.node, cur.opened)) {
                if cached >= cur.score {
                    continue;
                }
            }
            memo.insert((cur.node, cur.opened), cur.score);

            match result.get(&cur.opened) {
                Some(x) => result.insert(cur.opened, cur.score.max(*x)),
                None => result.insert(cur.opened, cur.score)
            };

            if !cur.opened[cur.node as usize] {
                let node = graph.get(&cur.node).unwrap();
                if node.0 > 0 {
                    let mut opened = cur.opened.clone();
                    opened[cur.node as usize] = true;
                    q.push_back(State { node: cur.node, score: &cur.score + (node.0 * (minute - 1)), opened });
                }
            }

            for next in &graph.get(&cur.node).unwrap().1 {
                q.push_back(State { node: *next, score: cur.score, opened: cur.opened });
            }
        }
    }

    result
}

fn part2() {
    let mut index = Index::new();
    let mut graph: HashMap<i64, (i64, Vec<i64>)> = HashMap::new();

    {
        let nodes: Vec<Node> = read_lines().iter().map(|x| parse(x)).collect();
        for node in nodes {
            graph.insert(index.get(node.from), (node.rate, node.tos.iter().map(|x| index.get(x.clone())).collect()));
        }
    }

    let start_idx = index.get("AA".to_string());
    let combinations = find_pairs(&graph, start_idx, 26);

    let mut result = 0i64;
    for comb in combinations {
        let elephant = bfs(&graph, start_idx, 26, comb.0);
        result = result.max(comb.1+elephant);
    }

    println!("Part2: {}", result);
}

fn main() {
    part1();
    part2();
}


